## the required packages ##
options(stringsAsFactors = F)

require(data.table)
require(tidyr)
require(vcd)
require(BEDMatrix)
require(snow)
require(dplyr)

### The original file processed by the following code is the association analyses result generated by the software regeine
### The suffix of output file produced by regenie is '.regenie'

################################# Single-variant tests (30 biochemistry markers) ##########################
## Process association analysis result files 
## This step is to filter significant variants with MAF<0.01
## Note:should change the path for differnt ethnic groups and different datasets
## three ethnicities: White,Asian,African
## five datasets：WGS(150k),TOPMed-imputed(150k),HRC+UK10K-imputed(150k),TOPMed-imputed(480k),HRC+UK10k-imputed(480k)
setwd("/home/biostat/DiskB/jldai/White/biochemistry_HRC")
threshold = -log10(5e-08)
for (pheno in 1:30) {
  biochemistry = data.table()
  for (chr in 1:22) {
    chr_biochemistry = fread(paste0("chr",chr,"_C",pheno,".regenie"))
    chr_biochemistry$maf = ifelse(chr_biochemistry$A1FREQ<=0.5,chr_biochemistry$A1FREQ,1-chr_biochemistry$A1FREQ)
    chr_biochemistry_keep = chr_biochemistry[chr_biochemistry$LOG10P>threshold & chr_biochemistry$maf<0.01,]
    biochemistry = rbind(biochemistry,chr_biochemistry_keep)
  }
  write.table(biochemistry,file = paste0("result/0.01summary/C",pheno,".csv"),col.names = T, row.names = F,sep = ",")
}

## Count the number of significant variation(MAF<0.01) of each trait
## Note:should change the path for differnt ethnic groups and different datasets
single_sum = data.table()
for (pheno in 1:30) {
  sum = 0
  type = paste0("c",pheno)
  now = fread(paste0("result/0.01summary/C",pheno,".csv"))
  sum = nrow(now)
  temp = data.table(type,sum)
  single_sum = rbind(single_sum,temp)
}
write.table(single_sum,file = "result/0.01summary/single0.01_sum.csv",col.names = T,row.names = F,sep = ",")

################################# Gene-based tests (30 biochemistry markers) ##########################
## Process association analysis result files 
## This step is to filter significant unique gene 
## Note:should change the path for differnt ethnic groups and different datasets
## three ethnicities: White,Asian,African
## five datasets：WGS(150k),TOPMed-imputed(150k),HRC+UK10K-imputed(150k),TOPMed-imputed(480k),HRC+UK10k-imputed(480k)
setwd("/home/biostat/DiskB/jldai/White/biochemistry_HRC")
threshold1 = -log10(2.5e-06)
for (pheno in 1:30) {
  biochemistry = data.table()
  for (chr in 1:22) {
    chr_biochemistry = fread(paste0("genebase_chr",chr,"_C",pheno,".regenie"))
    chr_biochemistry_keep = chr_biochemistry[LOG10P>threshold1,]
    biochemistry = rbind(biochemistry,chr_biochemistry_keep)
  }
  write.table(biochemistry,file = paste0("result/0.01summary/genebase_C",pheno,".csv"),col.names = T, row.names = F,sep = ",")
}

## Only retain the rows with UPPER==0.01 in the gene-based tests
## then drop the duplicate genes and finally count the number of genes
genebase = NULL
for (i in 1:30) {
  now  = fread(paste0("genebase_C",i,".csv"))
  
  # get the gene name（GENE）
  genebase_detail = separate(now,ID,into = c("GENE","other"),sep = "\\(")
  genebase_detail = genebase_detail[,-2]
  
  # get the MASK column and UPPER column 
  genebase_detail = separate(genebase_detail,ALLELE1,into = c("MASK","zero","upper"),sep = "\\.")
  genebase_detail = unite(genebase_detail,"UPPER",c("zero","upper"),sep = ".")

  genebase_detail$UPPER = gsub("singleton.NA","singleton",genebase_detail$UPPER)
  genebase_detail = genebase_detail[genebase_detail$UPPER=="0.01" & genebase_detail$LOG10P>(-log10(2.5e-06)),c("GENE","UPPER","LOG10P")]
  
  n = length(unique(genebase_detail$GENE))
  trait = paste0("c",i)
  
  temp = data.table(trait,n)
  genebase = rbind(genebase,temp)
}
write.table(genebase,file = "result/0.01summary/gene_unique_sum.csv",col.names = T,row.names = F,sep = ",")


################################# Single-variant tests (15 complex diseases) ##########################
## Process association analysis result files 
## This step is to filter significant variants with MAF<0.01
## Note:should change the path for different datasets
## one ethnicity: White
## five datasets：WGS(150k),TOPMed-imputed(150k),HRC+UK10K-imputed(150k),TOPMed-imputed(480k),HRC+UK10k-imputed(480k)
setwd("/home/biostat/DiskB/jldai/White/complex_HRC")
pheno = c("E10","E11","E66","F32","I10",
          "I25","I50","J44","J45","K80",
          "BLCA","BRCA","NHL","PRAD","SKCM")
threshold = -log10(5e-08)
for (trait in pheno) {
  summary0.01 = data.table()
  for (chr in 1:22) {
    chr = fread(paste0("chr",chr,"_",trait,".regenie"))
    chr$maf = ifelse(chr$A1FREQ<=0.5,chr$A1FREQ,1-chr$A1FREQ)
    chr_keep = chr_keep[chr_keep$LOG10P>threshold & chr_keep$maf<0.01,]
    summary0.01 = rbind(summary0.01,chr_keep)
  }
  write.table(summary0.01,file = paste0("binary_result/0.01summary/",trait,".csv"),col.names = T, row.names = F,quote = F,sep = ",")
}

## Count the number of significant variation(MAF<0.01) of each trait
## Note:should change the path for different datasets
single_sum = data.table()
for (trait in pheno) {
  sum = 0
  type = trait
  now = fread(paste0("bianry_result/0.01summary/",trait,".csv"))
  sum = nrow(now)
  temp = data.table(type,sum)
  single_sum = rbind(single_sum,temp)
}
write.table(single_sum,file = "binary_result/0.01summary/single0.01_sum.csv",col.names = T,row.names = F,sep = ",")

